### DISCUSSION TITLE
My case for `null`

### DISCUSSION THREAD

**Discussion by Lordfirespeed** [toml-lang/toml#1040] [General]
## Motivating use-case 1
I come from the JavaScript ecosystem where we have a package called [config](https://www.npmjs.com/package/config).

I really like the pattern it implements. I end up setting up my projects like this:
```
.
├── config
│   ├── .gitignore
│   ├── default.ts
│   ├── local.ts
│   └── production.ts
├── package.json
└── src
    └── ...
```

Now, 
- `default.ts` contains a complete 'development' configuration, and is included in VCS.
- `production.ts` contains a partial configuration of **overwrites** that are applied conditional on the `NODE_ENV` environment variable, and is included in VCS.
- `local.ts` contains a partial configuration of **overwrites** that are always applied, but is not included in VCS.
- The `config` library will deep-merge the structures exported from these files in a consistent, predictable order. 

In practise, this is good for various reasons ...
- `default.ts` acts as living documentation for the overwrite keys I can choose to add to `local.ts`, `production.ts`.  
- I don't have to repeat myself or remember to copy things from `default.ts` to `production.ts` when I make changes.
- I can specify machine-specific configuration in `local.ts`  without affecting collaborators.
- I can store sensitive credentials in `local.ts` without exposing them via my VCS.

Notably, though, for this pattern to work, I sometimes need to use `null` or `undefined` values in `production.ts` in order to 'unset' values defined by `default.ts`.

So when I came to trying to implement this pattern for a Python project using `.toml` files in place of `.ts`, I was fairly dismayed when I found out **I can't have `null` values in `.toml` files**!

## Motivating use-case 2
Say I have an 'optional' value which I load from a `.toml` file. 
```toml
foo = "bar"  # foo can be omitted
baz = "qux"
```
As long as the defined value for `foo` is not null-ish, it is obvious to any contributor to my project simply by looking in the config file that they can amend the value of `foo` to have some effect on the project's behaviour.

But if the defined value for `foo` **is** null-ish, I can either
- Comment out the line
- Omit `foo` from the config file

Both of these make it considerably less obvious to an arbitrary individual that `foo` is something they can amend the value of.

## Other remarks
`null` and `undefined` are simply 'sentinel' values roughly meaning 'the absence of a value'. 
If we want to get really silly, we could say TOML already supports `null`-ish values ... 

### Strategy 1: 'flags'
```toml
foo_is_null = true
foo = "bar" 

baz_is_null = false
baz = "qux"
```
`foo` -> `null` because `foo_is_null`
`baz` -> `"qux"` because `not baz_is_null`

### Strategy 2: bring your own sentinel value
select an arbitrary value that is unlikely to ever show up in common use ...
```bash
$ openssl rand -base64 48
7HPoQeiEIN8cE+hgW71KoiWCYDGM9DwKcF1fCVGDF0jZQbn5iEeBzLxYR5gK0LkR
```
and use it to represent `null` values
```toml
foo = "7HPoQeiEIN8cE+hgW71KoiWCYDGM9DwKcF1fCVGDF0jZQbn5iEeBzLxYR5gK0LkR"
baz = "qux"
```
`foo` -> `null` because its value is exactly the chosen sentinel value
`baz` -> `"qux"` because its value is not the chosen sentinel value

### fin.
since `null`-ish values are clearly 100% supported already (/s), what difference does adding a `null` keyword make? it would save people the effort of running `openssl` or duplicating all the keys in their files :0)

I want to be arguing over `null`/`nil`/`none`/`undefined`, not about whether TOML needs `null`!

**Comment by johnchristopherjones** [comment]
Part of your first use case is perplexing. If you "unset a default", then you have no default value.

I believe what's implied is that there is, in fact, another layer of defaults below `default.ts`. Let's say you're merging like this:

```
[real_defaults] <- [defaults.ts] <- [local.ts]
```

You could describe what you're doing as either:
- setting a value in `local.ts` to "whatever is in `real_defaults`" by using a singular magic value; or
- conditionally bypassing `default.ts`.

In either case, it seems more like an artifact that emerges from your merge strategy. The simplest solution to me seems to be _don't set it in "`default.ts`" in the first place)_. It's also a really common convention to generate "default" configs with some options commented out to indicate that you can can override them, which you could do here with `default.ts`. I don't think commented-out options make their availability "considerably less obvious", but JSON doesn't support comments, so perhaps it is more surprising for people who don't expect comments to be valid in config files.

**Reply by Lordfirespeed** [reply]
you're wrong. there is no layer below default.

If you want to be pedantic, fine: you can interpret my case as having `real_defaults` where all values in that layer are null-ish. 

When I say 'unset a default', I mean 'set a value defined in default.ts to some null-ish value'. This makes sense, because **almost every programming language ever** has a runtime representation of `null` or similar; why should it be impossible to specify that value explicitly using `.toml`?

Say we are (hypothetically) designing a language for representing numeric values. Should we represent zero as `0` or should we disallow that, and interpret only the lack of a value as zero? 

You're right - using the current toml spec, not setting defaults at all is the only way of doing things.
However, I described in detail the reasons **I do not want to do that**.

Your particular opinion about the obviousness of commented-out values is of little relevance, because I disagree, and others do too; the prevalence of those people does not matter, only that they exist.

JSON does not support comments. `.ts` (typescript) does.

**Comment by Hibbie89** [comment]
I want to add a use case that goes in the direction of use-case 2, but extends it in my case a bit.
I am developing a Python library to control a series of scientific measurement instruments, which base on slightly different combinations of hardware, and even for one instrument it might have different communication connections attached or not.
In this case, I made the design choice to have non-existend ports be set to None for mainly two reasons.

1. It is plain readable, and easy to understand. A "None" port can not be interpreted other than "it is not there".
- While I could switch to another value like empty string or such, this is not the same, as also discussed a lot in other threads.
- Commenting out the line would work in my case, except I want to write changes in the configuration from the software to a config file, and I have not found a good solution to work with commenting/uncommenting lines in TOML in Python programmatically despite searching quite a bit. Especially when retaining comments made to the line, this gets really messy really easily.
- Deleting the line is no option, as in this case the configuration file is no longer "self-documented". You can argue that people should read documentation, but especially in stressful situations, the configuration file must speak for itself.

2. Type checking warns me if I do not ensure that "None" is catched before I pass it to a setting function, which would cause additional errors. For various reasons, the software should not stop, so this is really helpful.

In hindsight, my fault was not to check this before deciding for a config file format (which would have likely ment not going with TOML), and in general I am very happy with this format (more so than standard config files). It is just really inconvenient, especially as None (or Null-Types in general) are present in a lot of programming languages, especially in configuration contexts.

**Comment by God-damnit-all** [comment]
@Lordfirespeed I feel like maybe you overcomplicated your examples and made them too language-specific. Here's my attempt at what I think you're getting at.

@aaronblohowiak Right now, TOML works great for a single configuration file, but it falls short when configs start becoming layered (like overriding a global config with a local one).

As an example of just one use-case, think about an end-user's typical setup for dev work when using a tool you've published. In their global config, they might do something like turn on file logging:

```toml
# ~/.config/really-cool-dev-tool/global.toml
[logging]
path = "/var/log/myapp.log"
level = "debug"
```

But in environments controlled by a local config, they could decide that file shouldn't be written to. Unfortunately, there's currently no clean way for local configs to say "have this option go back to the built-in default". Omitting the key doesn't work because the merging process would just let the global value fall through.

While a custom workaround in the tool's code for flagging when something should revert back to default would get the job done (e.g. a magic string like "DEFAULT"), that's a hacky band-aid instead of an intuitive standard.

Sure, one could determine what the default is supposed to be and assign that, but that's incredibly inconvenient for the end-user. They'd have to go digging through the documentation (hopefully up-to-date & of the correct version) or hunt through the source code to find it.

Additionally, having no value at all (not even an empty string) is often what the default is, which can't even be represented right now.

More importantly, this solution is fragile: if an internal default is updated later, the end-user that simply wants to fall back to the default value now has to revise each relevant local config (assuming they're paying enough attention to patch notes to realize the change was made).

Put simply, having a null value provides a potential method to explicitly unset options like the ones above:

```toml
# ./local.toml
[logging]
path = null
level = null
```

While this isn't completely foolproof (the code might be checking if a value is undefined rather than nullish), it would still establish a uniform way of being able to do something like this.

**Reply by Lordfirespeed** [reply]
Yes, this is very much aligned with my use-case, i.e. we may have some machine-specific `global.toml`, some `project.toml` checked into VC and also a machine-specific `local.toml`. Need to be able to merge those and occasionally 'unset' values in `local.toml` that were set in `project.toml`. 
Thanks!

**Reply by ChristianSi** [reply]
@arp242: Any chance you might reconsider your stance on `null` in TOML? Realistically, adding it seems to be the only chance of ever putting an end to these discussions!

Moreover, I think the arguments in this discussion have some merit, and there's also the case of arrays that may sometimes include null/undefined-like values – a situation TOML currently cannot handle.
